components-java-sdk/src/main/java/io/dapr/components/domain/state/Sorting.java

```diff
@@ -0,0 +1,54 @@
+/*
+ * Copyright 2023 The Dapr Authors
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.dapr.components.domain.state;
+
+import dapr.proto.components.v1.State;
+
+import java.util.Arrays;
+import java.util.Map;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+/**
+ * Describes a sorting order to be performed by a Query.
+ *
+ * @param key The key that should be used for sorting.
+ * @param order The order that should be used.
+ */
+public record Sorting(String key, Order order) {
+
+  enum Order {
+    ASC(dapr.proto.components.v1.State.Sorting.Order.ASC),
```

Comment:
Just a minor concern about tying new enum values to the values of enums not under our control; should their underlying values change, that could be breaking changes to applications that rely on those values (e.g. by caching them).  An explicit switch-like conversion function may be better (which, at least for .NET, enables the compiler to warn if/when new values are added).

Reply:
The problem is that if those enum values change, so does the resulting serialization under the wire, as those enums are coming directly from the Protocol Buffers generated code. Having a hardcoded integer instead won't change that and might instead make for harder to catch bugs say, in the case of a breaking change in the protocol.

Reply:
I could take an different approach, but I want to make sure we are on the same page (on both SDKs) as to what that alternative entails for risks and maintainability.

Reply:
In .NET, when working with enums that span two different sets of APIs, we'd typically add a short switch-based conversion function (for each way, as needed):

```csharp
public enum SdkOrder
{
  Unknown = 0,
  Ascending,
  Descending
}

public static SdkOrder ToSdkOrder(Grpc.Order order)
{
  return order swiitch
  {
    Grpc.Order.Asc => SdkOrder.Ascending,
    Grpc.Order.Desc => SdkOrder.Descending,
    _ => throw ArgumentOutOfRangeException(nameof(order))
  };
}
```

This is similar to the pattern-matching in Java 17 you mentioned earlier.  The C# compiler will warn/error if not all cases are accounted for, which is handy for maintainability.  It also means that we don't couple underlying values in one layer of the API to lower levels.  It may not matter as much in this scenario, but I'm inclined to still use this pattern in the .NET SDK regardless.  (It should be mostly invisible to the developer, in any case.  Though, now I'll have to go back and ensure that I'm abiding by my own words in what I've already added to the SDK. :-) )

Reply:
> This is similar to the pattern-matching in Java 17 you mentioned earlier. The C# compiler will warn/error if not all cases are accounted for, which is handy for maintainability. 

Ahhhh! Got it.

Well, Java8 doesn't. I can check if Java17 does but it seems this is something that is also behind the `--enable-preview` flag, [since pattern matching itself is behind this flag](https://www.baeldung.com/java-switch-pattern-matching#1-covering-all-values).

I can change this code into something that uses switch so we are future-proofed but, the way java ecosystem move fro LTS version to LTS version, this future proof would be pointless. I do think it would lead to a more straightforward code.

Reply:
The advantage of your visitor pattern suggestion was that it doesn't actually require exposing an enumeration to the user, therefore making at least part of this discussion moot.  I've also updated my [transaction PR](https://github.com/dapr-sandbox/components-dotnet-sdk/pull/9) to use a similar pattern.